### PR TITLE
Enable STI on claims

### PR DESCRIPTION
**Pull request by stephenrichards**
This PR turns adds Single Table Inheritance to the Claim model, renaming it Claim::BaseClaim.  This is
in preparation for creating a Claim::AdvocateClaim and Claim::LitigatorClaim model.

In addition to the renaming and migration of the claims table, changes were made to the Assessment model to ensure that
only one can ever exist for a claim - see note at top of Assessment model for details on how to create and update assessments.

### CODE REVIEW

File: app/models/json_schema.rb [hunk 1]
```diff
@@ -3,7 +3,7 @@ class JsonSchema
   class << self
 
     def generate(json_template)
-      schema = JSON::SchemaGenerator.generate 'Claim for crown court defence - claim import', json_template
+      schema = File.read(File.join(Rails.root, 'config', 'claim_schema.json'))
```
Comment (jsugarman) [inline app/models/json_schema.rb:6]: the arg json_template is never used. as we discussed if we go down the hard coded json schema route this should probably be removed.

File: Gemfile [hunk 2]
```diff
@@ -71,6 +71,7 @@ group :development, :test do
   gem 'rspec-collection_matchers'
   gem 'webrick',        '~> 1.3'
   gem 'parallel_tests'
+  gem 'literate_randomizer'
```
Comment (himal-mandalia) [inline Gemfile]: Faker Lorem module can be used for this:

http://www.rubydoc.info/github/stympy/faker/Faker/Lorem

File: app/models/claim/advocate_claim.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+#  original_submission_date :datetime
+#  retrial_started_at       :date
+#  retrial_estimated_length :integer          default(0)
+#  retrial_actual_length    :integer          default(0)
+#  retrial_concluded_at     :date
+#  type                     :string
+#
+
+module Claim
+	class AdvocateClaim < BaseClaim
```
Comment (himal-mandalia) [inline app/models/claim/advocate_claim.rb]: Indentation is 8 spaces?????

File: app/models/claim/base_claim.rb [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+#  original_submission_date :datetime
+#  retrial_started_at       :date
+#  retrial_estimated_length :integer          default(0)
+#  retrial_actual_length    :integer          default(0)
+#  retrial_concluded_at     :date
+#  type                     :string
+#
+
+module Claim
+	class BaseClaim < ActiveRecord::Base
```
Comment (himal-mandalia) [inline app/models/claim/base_claim.rb]: Indentation!

File: app/views/external_users/claims/_form.html.haml [hunk 5]
```diff
@@ -1,5 +1,6 @@
 #claim-form
-  = form_for [:external_users, @claim], builder: AdpFormBuilder, html: { novalidate: true }, multipart: true, authenticity_token: true do |f|
+  - url = @claim.persisted? ? external_users_claim_path(@claim) : external_users_claims_path
```
Comment (himal-mandalia) [inline app/views/external_users/claims/_form.html.haml]: Perhaps move this into the external users claims helper

File: app/views/shared/_message_controls.html.haml [hunk 6]
```diff
@@ -1,12 +1,12 @@
 = simple_form_for message, remote: true, authenticity_token: true  do |f|
```
Comment (himal-mandalia) [inline app/views/shared/_message_controls.html.haml:1]: This template is horrid. After we merge this PR we should refactor this.

File: db/migrate/20160208131812_add_type_to_claims.rb [hunk 7]
```diff
@@ -0,0 +1,10 @@
+class AddTypeToClaims < ActiveRecord::Migration
+  def up
+    add_column :claims, :type, :string
+    ActiveRecord::Base.connection.execute "UPDATE claims SET type = 'Claim::AdvocateClaim'"
+  end
+
+  def remove_column
+  	remove_column :claims, :type
```
Comment (himal-mandalia) [inline db/migrate/20160208131812_add_type_to_claims.rb]: Indentation

File: spec/factories/assessments.rb [hunk 8]
```diff
@@ -3,5 +3,10 @@
     claim         { FactoryGirl.create :claim, :without_assessment }
     fees          250.33
     expenses      845.89
+
+    trait :blank do
+    	fees				0
```
Comment (himal-mandalia) [inline spec/factories/assessments.rb]: Bring trailing 0 closer

File: spec/models/assessment_spec.rb [hunk 9]
```diff
@@ -87,4 +86,39 @@
...
+      assessment.update_values(888.88, 333.33, DateTime.new(2016, 1, 2, 3, 4, 5))
+      assessment.reload
+      expect(assessment.fees).to eq 888.88
+      expect(assessment.expenses).to eq 333.33
+      expect(assessment.created_at).to eq DateTime.new(2016, 1, 2, 3, 4, 5)
+    end
+  end
+
 end
+
```
Comment (himal-mandalia) [inline spec/models/assessment_spec.rb]: Trailing newlines

File: spec/presenters/base_claim_presenter_spec.rb [hunk 10]
```diff
@@ -79,22 +79,24 @@
...
         expect(subject.assessment_date).to eq '13/08/2015'
       end
     end
   end
 
-  describe 'assessment_fees' do
-    it 'should return formatted assessment fees' do
-      assessment = FactoryGirl.create :assessment, claim: claim, fees: 1234.56
+  describe 'asse
+  ssment_fees' do
```
Comment (himal-mandalia) [inline spec/presenters/base_claim_presenter_spec.rb]: Unnecessary newline

**Comment by jsugarman** [comment]
The api:smoke_test appears to be broken as a result of eternal user seeds

File: app/models/json_schema.rb (hunk 1 above)
Comment (stephenrichards) [inline app/models/json_schema.rb:6]: Ive raised a techdebt icket to look at this in more detail (https://www.pivotaltracker.com/story/show/113434493).  We need to take a decision as to whether or not to automatically generate the json schema as before or not, and then update the code accordingly.

File: Gemfile (hunk 2 above)
Comment (stephenrichards) [inline Gemfile]: Done.

File: app/models/claim/advocate_claim.rb (hunk 3 above)
Comment (stephenrichards) [inline app/models/claim/advocate_claim.rb]: fixed

File: app/models/claim/base_claim.rb (hunk 4 above)
Comment (stephenrichards) [inline app/models/claim/base_claim.rb]: fixed

File: app/views/shared/_message_controls.html.haml (hunk 6 above)
Comment (stephenrichards) [inline app/views/shared/_message_controls.html.haml:1]: have created a tech debt pivotal ticket for this. https://www.pivotaltracker.com/story/show/113732351

Comment (himal-mandalia) [inline app/views/shared/_message_controls.html.haml:1]: :thumbsup:

**Pull request merged**